clarity/src/vm/ast/parser_v2/mod.rs

```diff
@@ -0,0 +1,2518 @@
+                Err(e) => {
+                    assert!(
+                        fail_fast,
+                        "Parser::read_token should not return an error when not in fail_fast mode"
+                    );
+                    p.success = false;
+                    return Err(ParseErrors::Lexer(e));
+                }
+            };
+            if token.token == Token::Eof {
+                p.tokens.push(token);
+                break;
+            }
+            p.tokens.push(token);
+        }
+        p.diagnostics = p
+            .lexer
+            .diagnostics
+            .iter()
+            .map(|lex_error| PlacedError {
+                e: ParseErrors::Lexer(lex_error.e.clone()),
+                span: lex_error.span.clone(),
+            })
+            .collect();
+        p.success = p.lexer.success;
+        Ok(p)
+    }
+
+    fn add_diagnostic(&mut self, e: ParseErrors, span: Span) -> ParseResult<()> {
+        if self.fail_fast {
+            return Err(ParseError::new(e));
+        } else {
+            if e.level() == Level::Error {
+                self.success = false;
+            }
+            self.diagnostics.push(PlacedError { e, span });
+        }
+        Ok(())
+    }
+
+    fn next_token(&mut self) -> Option<PlacedToken> {
+        // Early exit in fail_fast mode.
+        if !self.success && self.fail_fast {
+            return None;
+        }
+
+        if self.next_token >= self.tokens.len() {
+            return None;
+        }
+        let token = self.tokens[self.next_token].clone();
+        self.next_token = self.next_token + 1;
+        Some(token)
+    }
+
+    fn ignore_whitespace(&mut self) -> bool {
+        let mut found = false;
+        loop {
+            if self.next_token >= self.tokens.len() {
+                return found;
+            }
+            let token = &self.tokens[self.next_token];
+            match &token.token {
+                Token::Whitespace => {
+                    self.next_token = self.next_token + 1;
+                    found = true;
+                }
+                Token::Comment(comment) => {
+                    self.comments.push(comment.clone());
+                    self.next_token = self.next_token + 1;
+                    found = true;
+                }
+                _ => return found,
+            }
+        }
+    }
+
+    fn parse_list(&mut self, lparen: PlacedToken) -> ParseResult<PreSymbolicExpression> {
+        let mut nodes = vec![];
+        let mut span = lparen.span.clone();
+        let mut whitespace = true;
+        loop {
+            if let Some(node) = self.parse_node()? {
+                if !whitespace {
+                    self.add_diagnostic(ParseErrors::ExpectedWhitespace, node.span.clone())?;
+                }
+                nodes.push(node);
+                whitespace = self.ignore_whitespace();
+            } else {
+                let token = self.tokens[self.next_token - 1].clone();
```

Comment:
We're sure that `self.next_token > 0` here, always?

Reply:
Yes, we know for sure that `self.next_token > 0` because the only way to get into `parse_list` is to have already processed a `(`, which would increment the count. I checked other locations where 1 is subtracted as well.

We also have a guarantee that there is always at least one token from the lexer, the `EOF` indicator, so even in the case of an empty string, one token will be processed successfully before reaching the end. This makes the `self.next_token - 1` in `parse_node_or_eof` safe.